src/ion/moves.rs

```diff
@@ -522,110 +620,61 @@ impl<'a, F: Function> Env<'a, F> {
-            let from_is_ret = self.func.is_ret(from_last_insn);
-            let to_is_entry = self.func.entry_block() == src.to_block();
-            let from_outs =
-                self.func.block_succs(src.from_block()).len() + if from_is_ret { 1 } else { 0 };
-            let to_ins =
-                self.func.block_preds(src.to_block()).len() + if to_is_entry { 1 } else { 0 };
-
-            let (insertion_point, prio) = if to_ins > 1 && from_outs <= 1 {
-                (
-                    // N.B.: though semantically the edge moves happen
-                    // after the branch, we must insert them before
-                    // the branch because otherwise, of course, they
-                    // would never execute. This is correct even in
-                    // the presence of branches that read register
-                    // inputs (e.g. conditional branches on some RISCs
-                    // that branch on reg zero/not-zero, or any
-                    // indirect branch), but for a very subtle reason:
-                    // all cases of such branches will (or should)
-                    // have multiple successors, and thus due to
-                    // critical-edge splitting, their successors will
-                    // have only the single predecessor, and we prefer
-                    // to insert at the head of the successor in that
-                    // case (rather than here). We make this a
-                    // requirement, in fact: the user of this library
-                    // shall not read registers in a branch
-                    // instruction of there is only one successor per
-                    // the given CFG information.
-                    ProgPoint::before(from_last_insn),
-                    InsertMovePrio::OutEdgeMoves,
-                )
-            } else if to_ins <= 1 {
-                (
-                    ProgPoint::before(to_first_insn),
-                    InsertMovePrio::InEdgeMoves,
-                )
-            } else {
-                panic!(
-                    "Critical edge: can't insert moves between blocks {:?} and {:?}",
-                    src.from_block(),
-                    src.to_block()
-                );
-            };
+        if !block_param_dests.is_empty() {
+            // Sort the sources to mirror the destinations now, ordering by dest vreg/dest
+            // block/source block.
+            block_param_sources.sort_unstable_by_key(|h| h.key());
 
-            let mut last = None;
-            for dest in first_dest..last_dest {
-                let dest = &half_moves[dest];
-                if last == Some(dest.alloc) {
-                    continue;
+            let mut src_ix = 0;
+            trace!("processing block-param moves");
+            for dest in block_param_dests {
+                // Find the source for the current destination
+                loop {
```

Comment:
I think we need a `while src_idx < ...` here? We might be OK for now because we'll always find something but this would make for an unfortunate debugging session later if that changes :-)

Reply:
Also, let's add a note here about why we expect the same order in `block_param_dests` and `block_param_sources`, namely that we depend on a correspondence between the key functions and sorts of both.

Reply:
I refactored this to use a peekable iterator instead of a second index, and added a comment that explains the order invariants of those two vectors.